[#74] Feat: 짤 이미지 신고 api 연결 구현

HY219 opened the pull request:
## 📝 작업 내용

- 이미지 상세 모달에서 `신고하기 이미지 버튼` 클릭 시 `신고하기 POST API`를 요청합니다.
- `/v1/report/{$imageId}` 로 POST 요청

+`handleClickReportButton` 이름이 한 파일에서 중복으로 쓰여 문제가 발생해, 신고 상세 모달에서의 핸들러 함수 이름은 `handleClickReportCompleteButton`으로 수정해주었습니다

+코드를 작성한 `ImageMenuBar.tsx` 파일이 다른 파일과 통합되면서 main에서 사용되지 않아, 추후 통합이 필요합니다.

## 📷 스크린샷 (선택)

## 💬 리뷰 요구사항(선택)

- **`./src/components/ImageDetailModal/ImageMenuBar.tsx` 파일의 `13번째 줄`과
`./src/apis/zzal.ts` 파일의 `4번째 줄`의
 `변수명`이 `postReportZzal` 으로 같아 그대로 두어도 상관없는지 변경하면 좋을지 궁금합니다!**
  - **변경해야한다면, 코드의 통일성을 위해서 변수명을 어떻게 작성할지 정하면 좋을 것 같습니다!**

  `./src/components/ImageDetailModal/ImageMenuBar.tsx`
  ![image](https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/198b1f59-aee1-4e6b-84a7-71c8acb5c015)

  `./src/apis/zzal.ts`
  ![image](https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/b40f4b85-f0e9-4ce7-bf8c-cdb3f13b949c)

- **`신고하기 POST API를 호출하는 함수와 타입 정의`를 `zzal.ts`, `zzal.dto.ts` 파일에 작성해주었는데, `report.ts`, `report.dto.ts` 파일에 작성하는 것이 더 좋을지 궁금합니다!**
  - **짤 상세 모달 메뉴에 위치해 있는 `좋아요`, `다운로드`등의 기능들이 `zzal.ts` 에 포함이 될 것이라 생각해 `zzal.ts`에 코드를 작성했습니다.**
  - **또 현재 `.src/types/report.ts` 파일에 정의 해둔 `인터페이스명`과 헷갈릴 여지가 있어보여 `zzal.ts`에 코드를 작성했습니다.**

- 변수명, 함수명, 폴더명이 가독성있게 작성되었는지 궁금합니다!


## 📍 기타 (선택)

- 이미지 상세보기 api 연결 후, 실제  imageId 추가 필요
  - `TODO: [2024-02-28] 이미지 상세보기 api 연결 후, 실제 imageId를 가져와야합니다.`
- 이미지 상세 모달에서 `신고하기 이미지 버튼` 클릭 시 `신고 확인 모달` 을 먼저 출력해야하는데, 새로 이슈 생성해서 구현하도록 하겠습니다:)

- 추가로, 이미 이미지를 한 번 신고한 유저가 같은 이미지를 재신고하는 동작은 허용되지 않도록 서버에서 작업해주신 것 같아요.
![image](https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/c1f330e1-13ff-4be2-beae-ee49f6af0397)


  =>  **신고했던 이미지를 재신고하는 동작을 했을 때, '이미 신고한 이미지입니다.'라는 `toast 메시지` 또는 `모달`을 띄우면 좋을 것 같아요!**

  **정하면 좋을 것 같습니다:)**

- 코드 동작을 브라우저에서 확인하려면, `./src/routes/_layout-with-chat/index.tsx` 파일에  `ImageDetailModal 컴포넌트`를 import 하고, `4번째 줄`에 다음 코드를 추가해 주세요:)
  > return ( <><ImageDetailModal isOpen={true} onClose={() => {}} /></> );

close #74

HY219 committed:
Feat: 짤 이미지 신고 api 연결 구현

짤 이미지 상세 모달에서 신고 이미지 버튼 클릭 시 POST 요청
/report/ 로 POST 요청
이미지 상세보기 api 연결 구현 후, 실제 imageId 추가 필요

yjc2021 commented on src/apis/zzal.ts:
```diff
@@ -0,0 +1,8 @@
+import { PostReportZzalRequeust } from "@/types/zzal.dto";
+import http from "./core";
+
+export const postReportZzal = ({ imageId }: PostReportZzalRequeust) => {
```
오타가 있네요! `imageId`로 수정 부탁드립니다 👍

HY219 replied:
엇 혹시 일치해보이는데 어느 부분일까요? 👀

thalassophilia commented on src/types/zzal.dto.ts:
report는 말씀해주신대로 report.dto.ts로 분리하시면 더 좋을 것 같습니다!

thalassophilia commented on src/hooks/api/zzal/usePostReportZzal.ts:
```diff
@@ -0,0 +1,14 @@
+import { postReportZzal } from "@/apis/zzal";
+
+const usePostReportZzal = () => {
+  const mutation = useMutation({
```
mutate 함수만 이름 재정의해서 내보내주시면 더 좋을 것 같은데 어떤가요?

thalassophilia commented on src/hooks/api/zzal/usePostReportZzal.ts:
```diff
@@ -0,0 +1,14 @@
+const usePostReportZzal = () => {
+  const mutation = useMutation({
+    mutationFn: postReportZzal,
+    onSuccess: () => console.log("postReportZzal success"),
+    onError: () => console.log("postReportZzal error"),
```
불필요한 로깅 같아서 삭제해주시면 좋을 것 같습니다!

thalassophilia commented on src/components/ImageDetailModal/ImageMenuBar.tsx:
```diff
@@ -1,18 +1,28 @@
+
+  const handleReportZzal = () => {
+    postReportZzal.mutate({
+      imageId: imageId,
```
imageId로 축약해서 쓸 수 있을 것 같습니다!

thalassophilia commented on src/apis/zzal.ts:
신고는 zzal 도메인과는 다른 영역 같아서 새로 파일 생성해주시면 더 좋을 것 같아요 : )

thalassophilia commented on src/apis/zzal.ts:
```diff
@@ -0,0 +1,8 @@
+import http from "./core";
+
+export const postReportZzal = ({ imageId }: PostReportZzalRequeust) => {
+  return http.post<PostReportZzalRequeust>({
```
response 타입을 지정해주는 자리인데 이름이 `PostReportZzalRequeust`라서 좀 헷갈리는 것 같아요! 타입이 같더라도 response 타입을 생성해주시면 좋을 것 같습니다 : )

HY219 committed:
Rename: zzal 폴더 내 신고 관련 api 요청 데이터 전송 객체를 report폴더로 이동

HY219 committed:
Refactor: mutate 함수 재정의 및 요청 파라미터 축약,  불필요한 로깅 삭제

POST 요청이 성공/실패 했을 때 동작하는 로깅 삭제
요청 파라미터 imageId 축약
재신고 시 POST 요청이 실패 되므로, 추후 POST 요청 실패 시 알림 코드 추가 필요

HY219 committed:
Refactor: POST 요청에 대한 응답 타입 삭제 및 비동기 처리 추가

POST 요청에 대한 응답 타입이 존재하지 않으므로 타입 삭제
POST 요청에 비동기 처리를 하여 응답 대기 시간 단축

thalassophilia commented on src/apis/report.ts:
```diff
@@ -0,0 +1,10 @@
+import http from "./core";
+
+export const postReportZzal = async ({ imageId }: PostReportZzalRequeust) => {
+  const response = await http.post({
+    url: `/report/${imageId}`,
+  });
+
+  return response;
```
response 생성하지 마시고 바로 반환해주시면 될 것 같습니다!
async, await도 없어도 될 것 같아요~

HY219 replied:
신고 요청은 바로 처리가 되기 때문에 비동기 처리가 따로 필요없는걸까요?
정확히 몰라 궁금해서 여쭤봅니다!

choi-ik replied:
mutationFn은 내부적으로 비동기 함수를 지원해서 promise를 반환해요! 그래서 사용하는 함수가 promise를 반환하지 않을 때만 async/await 키워들 붙혀주면 됩니다 ㅎㅎ

또한 지금 저희는 axios를 사용하고 있고, fetch와 다르게 상태코드가 400일때 reject이 됩니다! 그래서 reject 되었을 때의 async/await에 대한 후처리도 현 코드에서는 필요가 없죠 ㅎㅎ (mutationFn과 queryFn이 error임을 나타내려면 reject 되었음을 인지해야 하는데 fetch는 reject이 아닌 resolve된 promise의 ok에 false로만 표기하는 걸로 알고있어요)

thalassophilia commented on src/hooks/api/zzal/usePostReportZzal.ts:
```diff
@@ -0,0 +1,12 @@
+    mutationFn: postReportZzal,
+  });
+
+  return reportZzal;
```
이렇게 내보내면 isPending이나 error는 사용 못해요ㅠㅠ아래처럼 수정해주세요
```js
const {mutate, ..rest} = useMutation()

return {
 reportZzal: mutate,
...rest
}
```

thalassophilia commented on src/components/ImageDetailModal/ImageMenuBar.tsx:
```diff
@@ -1,18 +1,28 @@
+  const reportZzal = usePostReportZzal();
+
+  const handleReportZzal = () => {
+    reportZzal({
+      imageId,
+    });
```
객체 형식이 아니어도 될 것 같아요!

thalassophilia commented on src/components/ImageDetailModal/ImageMenuBar.tsx:
```diff
@@ -1,18 +1,28 @@
 
+  const reportZzal = usePostReportZzal();
+
+  const handleReportZzal = () => {
```
`handleReportZzal`은 컨벤션에 맞지 않은 이름 같습니다! 신고하기 버튼을 눌렀을때 호출되는 함수이니 `handleClickReportButton`이 더 적절한 것 같은데 어떠세요?

HY219 replied:
좋습니다! 수정하겠습니다~!

thalassophilia commented on src/apis/report.ts:
```diff
@@ -0,0 +1,10 @@
+import http from "./core";
+
+export const postReportZzal = async ({ imageId }: PostReportZzalRequeust) => {
+  const response = await http.post({
```
response 타입 명시해주세요ㅠㅠㅠㅠ아래처럼 response 타입이 들어가야 하는데 빠져있네요
```js
http.post<PostReportZzalResponse>()
```

HY219 replied:
POST 요청을 했을 때 응답 데이터가 따로 없는 것 같은데 혹시 어떻게 명시를 해줘야하나요?
![image](https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/c131da33-7b90-48bd-878c-472ca3421793)

choi-ik replied:
```javascript
http.post<void>()
```
요런식으로 해주면 될 것 같네요??

HY219 committed:
Refactor: api 요청 response 타입 명시

response 데이터가 없어서, void로 명시

HY219 committed:
Rename: 신고하기 버튼을 클릭했을 때 호출되는 함수명 변경

handleReportZzal -> handleClickReportButton

HY219 committed:
Refactor: api 요청 파라미터를 객체에서 변수 형식으로 변경

객체 형식으로 받아오던 imageId를 변수 형식으로 변경
객체 형식 타입 정의 제거

HY219 committed:
Refactor: 추가적인 mutation 옵션을 사용할 수 있도록 나머지 연산자 추가

isSuccess, isError 등 추가 option 사용 가능

HY219 committed:
Refactor: API 요청에서 비동기 처리 제거 및 요청 결과를 직접 반환하도록 수정

MutateFunction이 Promise를 반환해줌으로 async/await 처리를 해주지 않아도 된다.

thalassophilia commented on src/apis/report.ts:
```diff
@@ -0,0 +1,5 @@
+import http from "./core";
+
+export const postReportZzal = (imageId: number) => {
+  return http.post<void>({ url: `/report/${imageId}` });
```
`v1`이 빠져있는거 같은데 추가해주세요!

HY219 committed:
Fix: api 요청 url 수정

C:/Program Files/Git/report/ -> /v1/report/ 로 변경

HY219 committed:
Merge branch 'main' into #74/feature/report-zzal-api

hyeonjinan096 commented on src/apis/report.ts:
```diff
@@ -1,12 +1,16 @@
+
+export const postReportZzal = (imageId: number) => {
+  return http.post<void>({ url: `/v1/report/${imageId}` });
+};
```
RportedImage랑 postReportZzal에서 Image랑 Zzal 중에 하나로 통일해야할 것 같네요:) 전체적으로 머지되면 Zzal로 다 통일할까요?

thalassophilia replied:
오 좋은 의견이네요! 신고된 이미지 삭제 pr 지금 살아있으면 바로 수정하면 될 거 같네요!

HY219 replied:
좋습니다~ 수정해서 반영하도록 하겠습니다!

thalassophilia commented on src/components/ImageDetailModal/ImageMenuBar.tsx:
```diff
@@ -1,18 +1,26 @@
     setIsLiked((prevLiked) => !prevLiked);
   };
 
+  const { reportZzal } = usePostReportZzal();
```
api 커스텀 훅 선언문은 상위에 선언해주시면 좋을 것 같네요!

yjc2021 commented on src/hooks/api/zzal/usePostReportZzal.ts:
```diff
@@ -0,0 +1,12 @@
+
+const usePostReportZzal = () => {
+  const { mutate, ...rest } = useMutation({
+    mutationFn: postReportZzal,
```
onSuccess, onError 등의 로직이 이후 추가되어야 할 것 같아서 TODO 주석 달아주시면 감사하겠습니다

HY219 replied:
onSuccess에 "신고가 접수되었습니다"라는 toast 띄우도록 하겠습니다!

onError 에는 error가 500일 경우와 400일 경우 다른 toast 메시지를 띄워야 할 것 같아서 좀 더 알아보도록 하겠습니다:)

HY219 replied:
onError 토스트 추가했는데, 더 좋은 방법이 있다면 알려주세요!

HY219 replied:
axios 에서 오는 error response status를 사용하여 코드를 구현해서, 제거해주었습니다.

현재 필요한 http error code는 서버 커스텀 error이기 때문에, 새로 이슈 생성해서 작업이 필요합니다.

TODO로 작성해 두었습니다:)
>  TODO: [2024-03-06] http error code 별 메세지(ex. 이미 신고가 완료되었습니다) 추가

HY219 committed:
Refactor: api 커스텀 훅인 usePostReportZzal() 선언문의 위치 상위로 변경

HY219 committed:
Merge branch 'main' of https://github.com/zzalmyu/zzalmyu-frontend into #74/feature/report-zzal-api

HY219 committed:
Feat: 짤 이미지 신고 api 성공 시 toast 구현

HY219 committed:
Rename: DeleteReportImage를 DeleteReportZzal로 변경

통일성을 위해 변경

thalassophilia reviewed (approved):
고생하셨습니다!

choi-ik commented on src/components/ImageDetailModal/ImageMenuBar.tsx:
```diff
@@ -1,18 +1,30 @@
+  const imageId = 152; // TODO: [2024-02-28] 이미지 상세보기 api 연결 후, 실제 imageId를 가져와야합니다.
   const [isLiked, setIsLiked] = useState(false);
 
   const handleClickLike = () => {
```
이 부분도 토글 버튼이라는 걸 나타내는 네이밍이 들어가면 좋을 것 같아요!

HY219 replied:
이 부분도 같은 부분인데, 현재 PR이 좋아요와 관련 없는 부분이라 현재 PR에서 수정하는게 맞을까요?

choi-ik replied:
아 그런거면 아직 수정하지 않아도 될 것 같아요! 좀 더 다른 분들과 이야기 해보고 고쳐보죠 ㅎㅎ

choi-ik commented on src/components/ImageDetailModal/ImageMenuBar.tsx:
```diff
@@ -1,18 +1,30 @@
   };
 
+  const handleClickReportButton = () => {
+    reportZzal(imageId, {
+      onSuccess: () => {
+        toast.success("신고가 완료되었습니다.");
+      },
+    });
```
이 부분 mutation 선언 한 파일에서 적어도 되지 않을까요??

HY219 replied:
api 성공/실패 시 toast와 함께 다른 작업을 수행해야 할 경우가 있는데, toast만 있는 경우에는 mutation을 선언한 파일에 onSuccess/onError를 작성해주고, 그렇지 않은 경우에는 파일 커스텀 훅 api를 사용한 파일에 작성하는 것이 가독성이나 통일성면에서 좋은지 모르겠습니다!

혹시 의견이 다르시다면 말씀해주세요!

HY219 replied:
#68 PR에서 짤 업로드 api를 연결 후, 현재 파일처럼 커스텀 훅 api를 사용한 파일에 있는 값을 onSuccess 내에서 사용하기 위해 mutation 선언한 파일이 아닌 현재처럼 mutate를 사용한 해당 파일에 onSuccess, onError를 작성해주었었습니다.
리뷰를 반영하면서 #68 을 위와 같이 작성하였는데, 이 코드와의 통일성을 위해서 #76 를 작성했습니다.

choi-ik replied:
음 혜연님이 말씀하신 것 같은 경우에는 mutation이 return해주는 isError를 사용해서 처리를 해주어도 될 것 같아요!

HY219 replied:
이 부분 내일 혹시 설명 부탁드려도 될까요? 
다음 글을 봐서 isSuccess/isError 보다 onSuccess/onError를 사용하는 것이 더 적합하다고 생각한 것 같습니다!
재민님께서 전에 리뷰에서도 isSuccess가 아닌 onSuccess를 사용하는 코드를 리뷰로 주셨어서, 이유를 여쭤보고 싶네요:)

이 둘의 차이에 대해서 자료를 찾고 싶은데 쉽지가 않아서, 혹시 아신다면 부탁드려보겠습니다 :)
![image](https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/8e12c9f2-9f49-48ea-8208-cadfaee8ebab)
https://stackoverflow.com/questions/77077506/diffrence-between-issuccess-and-onsuccess-in-react-query

yjc2021 replied:
useMutate()과 mutate() 에서 사용되는 onSucess의 차이점에 대한 아티클입니다 참고 부탁드립니다

[Some callbacks might not fire](https://tkdodo.eu/blog/mastering-mutations-in-react-query#some-callbacks-might-not-fire)

yjc2021 replied:
권장사항은 query관련 로직은 useMutation의 onSuccess, UI관련 로직 (toast, navigation)은 mutate의 onSucess에 명세하는 것입니다. 
useMutation의 onSuccess는 실행이 보장되지만, mutate의 onSuccess는 mutation 완료 전에 mutate을 호출하는 컴포넌트가 unmount되면 실행되지 않습니다
따라서 혜연님이 작성하신 대로 mutate()의 options에서 onSuccess를 명세하는 게 더 좋은 practice 같습니다

choi-ik replied:
오 이건 몰랐네요 대박쓰 👍🏻👍🏻👍🏻

HY219 committed:
Feat: 신고 요청 api 실패 시 에러 코드에 따라 toast 메시지 출력

400: 이미 신고가 완료되었습니다.
500: 신고가 올바르게 되지 않았습니다.

HY219 committed:
Merge branch 'main' of https://github.com/zzalmyu/zzalmyu-frontend into #74/feature/report-zzal-api

HY219 committed:
Merge branch 'main' of https://github.com/zzalmyu/zzalmyu-frontend into #74/feature/report-zzal-api

HY219 committed:
Cleanup: 짤 이미지 신고 api 연결 및 해당 TODO 주석 삭제

HY219 committed:
Fix: 신고 확인 모달에서 신고 버튼 클릭 시 http error code 별 메시지 출력 코드 제거

추후 작업
> 기존에 작성했던 axios errorcode와 실제 필요한 error 코드인 서버의 custom error는 다르다
TODO: 이미 신고한 짤일 경우 알맞는 Toast메시지 출력
TODO: 서버 커스텀 http error code 별 Toast메시지 출력
TODO: http 요청이 예상치 못하게 실패했을 경우 Toast메시지 출력

yjc2021 reviewed (approved):
수고하셨습니다 LGTM 👍

choi-ik reviewed (approved):
LGTM!!!🚀🚀🚀🚀🚀🚀🚀🚀🚀 x 5조 5억

hyeonjinan096 reviewed (approved):
수고하셨습니다👍👍👍
